### PR TITLE
[WiP/draft] Sub-domain world generators

### PR DESCRIPTION
**Pull request by ibrahimkakbar**
**Basics**: `world_generator_cave.cc` extended to the urban/tunnel sub-domains
_Note_: [Tunnel domain PR](https://github.com/osrf/subt/pull/720) merge is to the dome branch and not master; I include the code base for the `tunnel_generator` but don't build and install it in the CMakeLists.txt; I have not yet verified if the tunnel tiles  positions are correct.

**Usage**: urban_generator [world_type] [options]
Example Generate: urban_generator -g -t  s -c 60 -n urban_test -s 25 -o urban_test.sdf
Example Debug: urban_generator -g -d "Urban Straight" -n urban_straight -o urban_straight.sdf
Options:
    -h		 Print this help message
    -d <tile>	 Generate world from tile for debugging
    -o <file>	 Output sdf filename
    -t <type>	 Urban Type:
             	    'subways' or 's',
             	    'buildings' or 'b',
             	    'mixed' or 'm'
    -s <seed>	 Seed
    -c <count>	 Min tile count
    -n <name>	 World name
    -g		 Generate sdf with GUI plugin

Substitute [urban]_generator with: cave, tunnel for their respective domains.
- [x] Generate cave world
- [x] Generate cave debugging world
- [ ] Generate urban world **SystemPath.cc/Orientation Error**
  - [ ] Generate urban subway world
  - [ ] Generate urban building world
  - [ ] Generate urban mixed structure world
- [ ] Generate urban debugging world
  - [ ] Debug building tiles
  - [ ] Debug subway tiles
  - [ ] Connect mixed tiles
- [ ] Generate tunnel world **SystemPath.cc/Orientation Error** 
- [ ] Generate tunnel debugging world

### Cave Generator Environment
cave_generator

### Initial Urban Generator Environment
urban_generator

### CODE REVIEW

File: subt_ign/CMakeLists.txt [hunk 1]
```diff
@@ -4,7 +4,7 @@ project(subt_ign)
 configure_file (${PROJECT_SOURCE_DIR}/include/subt_ign/Config.hh.in ${PROJECT_BINARY_DIR}/include/subt_ign/Config.hh)
 
 if(NOT WIN32)
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17 -Wall -Wextra")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17 -Wall -Wextra -g")
```
Comment (nkoenig) [inline subt_ign/CMakeLists.txt:7]: Probably don't want the `-g` option here.

**Review by nkoenig — COMMENTED** [review]
It looks like a nontrivial portion of this code has been copied between files. I'm not sure what is new code and what is old code. 

Would it be possible to submit a new PR that just consists of code copy? Then this PR can show diffs.

**Comment by ibrahimkakbar** [comment]
@nkoenig I removed the old code from this PR to show the diffs.

**Comment by nkoenig** [comment]
@ibrahimkakbar , about how long is the `urban_generator` expected to run for? I'm trying `urban_generator -g -t -c 60 -n urban_test -s 25 -o urban_test.sdf` and it's been running for over 10 minutes.

**Comment by ibrahimkakbar** [comment]
@nkoenig The input is missing the type specification `urban_generator -g -t s -c 60 -n urban_test -s 25 -o urban_test.sdf`

**Comment by nkoenig** [comment]
Thanks. That works for me. 
This compiles and runs, and the code looks okay. How about we merge this PR and tackle the remaining work in additional PRs?

**Comment by ibrahimkakbar** [comment]
That works for me I can open separate PRs for the orientation issues per sub-domain and wiki update.

**Pull request merged**